### ISSUE TITLE
about loss function to train critic

### ISSUE THREAD

**Issue by CDLCHOI** [ou524u/MotionCritic#9]
Hi, 
In your paper, you use sigmoid() to train critic model,
But in code, MotionCritic/train.py, you use a cross-entropy with the label 0 to constrain the output prob of good sample close to value 1.
Does that mean cross-entropy is better than sigmoid for training critic model?

**Comment by ou524u** [comment]
Thanks a lot for the question and for checking both the paper and the code in detail.

In the paper, we write the critic loss in a logistic form $L_{\text{percept}} = - \mathbb{E}_{(x^{(h)}, x^{(l)})} \log \sigma\big( C(x^{(h)}) - C(x^{(l)}) \big)$, which is a standard pairwise logistic loss on the score difference.

In the code, we implement the critic as a 2-logit classifier and use cross-entropy with label 0:

```python
def loss_func(critic):
    target = torch.zeros(critic.shape[0], dtype=torch.long).to(critic.device)
    loss_list = torch.nn.functional.cross_entropy(critic, target, reduction='none')
    loss = torch.mean(loss_list)

    critic_diff = critic[:, 0] - critic[:, 1]
    acc = torch.mean((critic_diff > 0).clone().detach().float())
    return loss, loss_list, acc
```

If we denote the two logits by $z_0$, $z_1$, then 

$\text{CE}([z_0, z_1], y=0)  = -\log \frac{e^{z_0}}{e^{z_0} + e^{z_1}}  = -\log \sigma(z_0 - z_1).$

So the 2-way cross-entropy in the code is mathematically equivalent to the $-\log\sigma(\cdot)$ loss in the paper (applied to the logit difference). It’s just an equivalent parametrization of the same logistic loss, not a different objective.
